terraform/modules/scheduled-function/main.tf

```diff
@@ -0,0 +1,169 @@
+# Scheduled Cloud Function Module
+# This module creates a complete scheduled Cloud Function setup with all necessary components
+
+# Required providers
+terraform {
+  required_version = ">= 1.3.0"
+
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = ">= 6.0.0"
+    }
+    archive = {
+      source  = "hashicorp/archive"
+      version = ">= 2.0.0"
+    }
+    null = {
+      source  = "hashicorp/null"
+      version = ">= 3.0.0"
+    }
+  }
+}
+
+# Service account for the Cloud Function
+resource "google_service_account" "function_sa" {
+  project      = var.project_id
+  account_id   = "${var.function_name}-sa"
+  display_name = "Service Account for ${var.function_name} function"
+  description  = "Service account used by the ${var.function_name} scheduled Cloud Function"
+}
+
+# Storage bucket for function source code
+resource "google_storage_bucket" "function_bucket" {
+  project                     = var.project_id
+  name                        = "${var.function_name}-source-${var.project_id}"
+  location                    = var.region
+  uniform_bucket_level_access = true
+
+  # Optional: Add lifecycle rules to clean up old versions
```

Comment:
I'm not sure I understand this.  Terraform will only upload code if it's changed right?  So for code that never changes, this will delete the code after a week, and then terraform will have to re-upload it, is that right?

Reply:
An alternative is to always just name the tarfile with a fixed name, and set up versioned objects on the gcs bucket, and then have a lifecycle rule to delete all versions except the most recent 3 (or whatever).